src/rafimage.cpp

```diff
@@ -248,17 +249,31 @@ void RafImage::readMetadata() {
 
   Internal::enforce(jpg_img_len >= 12, ErrorCode::kerCorruptedMetadata);
 
-  DataBuf buf(jpg_img_len - 12);
-  if (io_->seek(jpg_img_off + 12, BasicIo::beg) != 0)
+  DataBuf jpg_buf(jpg_img_len);
+  if (io_->seek(jpg_img_off, BasicIo::beg) != 0)
     throw Error(ErrorCode::kerFailedToReadImageData);
 
-  if (!buf.empty()) {
-    io_->read(buf.data(), buf.size());
+  if (!jpg_buf.empty()) {
+    io_->read(jpg_buf.data(), jpg_buf.size());
     if (io_->error() || io_->eof())
       throw Error(ErrorCode::kerFailedToReadImageData);
   }
 
-  ByteOrder bo = TiffParser::decode(exifData_, iptcData_, xmpData_, buf.c_data(), buf.size());
+  // Extracting metadata from first APP1 container
+  constexpr byte tiff_offset = 12;
+  ByteOrder bo = TiffParser::decode(exifData_, iptcData_, xmpData_, jpg_buf.c_data() + tiff_offset, jpg_buf.size() - tiff_offset);
+
+  // If there is no XMP data in first APP1 EXIF structure, there is maybe
+  // another APP1 container for XMP. Just use the JpegImage metadata parser for that.
+  if(xmpData_.empty()) {
+    BasicIo::UniquePtr jpg_io = std::make_unique<Exiv2::MemIo>(jpg_buf.data(), jpg_buf.size());
+    auto jpg_img = JpegImage(std::move(jpg_io), false);
```

Comment:
Why not use this single decoder to extract both Exif and XMP in one shot?

I guess it might not be guaranteed that the above Exif APP1 segment assumed to be handled by TiffParser comes first...

Reply:
@kmilos Yes, thats my guess, too. And I don't know how I should get the byte order for `setByteOrder(bo);`.

Reply:
You don't need it or care about it here. The JpegImage figures it out internally anyway.

Reply:
I can maybe dig deeper into this, but not sure why `setByteOrder(bo)` is there to begin with... There is parsing of another Tiff structure in the RAF later, and that has it's own byte order (we don't know if it's guaranteed to be identical to the embedded JPEG one, but it most likely is given that it is serialized by the same firmware).

If you want to keep the same functionality, then you can just do something like `setByteOrder(jpg_img.byteOrder())` instead.